app/grandchallenge/reader_studies/views.py

```diff
@@ -1409,14 +1409,15 @@ def form_valid(self, form):
         instance = self.get_object()
         assigned_civs = []
         for ci_slug, new_value in form.cleaned_data.items():
-            if ci_slug == "order" or new_value is None:
+            if ci_slug == "order":
                 continue
-            instance, assigned_civs = self.create_civ_for_value(
-                instance=instance,
-                ci_slug=ci_slug,
-                new_value=new_value,
-                assigned_civs=assigned_civs,
-            )
+            with transaction.atomic():
```

Comment:
Why is this needed? We should be using atomic requests.

Reply:
Ok, brainfog on Monday evening. It turns out to be unnecessary here, just tested things again. I think what happened is that I misplaced this on Monday evening, because it does appear to be necessary in the `DisplaySetInterfacesCreate` view, but I don't understand why. We call `create_instance` there and I see in backends/base.py that you also wrapped the output interface creation in an atomic block, explaining there that this is necessary because `create_instance` needs to create a CIV to store the files. I don't follow this explanation though. We create CIVs in a lot of places where the extra atomic block is not necessary, so why with this specific method then?

Reply:
The only thing I can think of is that those are used in non-atomic Celery tasks, in which case the atomicity should be moved into the task itself (sometimes we don't want a task to be completely atomic, e.g. with image imports where you want to inform the user that something is being consumed first then a large amount of work happens). You have to be really careful with refactoring those. I just pointed it out here as there shouldn't be a reason to introduce any new ones outside of a Celery task.

There was a time when we were not using atomic commits due to an error on my part (put the setting in the wrong place for years!) and there might be remnants of that, but it is not a cause of the issue you are trying to fix.

Reply:
While trying to write a test to illustrate the issue in the `DisplaySetInterfacesCreate`, I realized that the atomic block is only necessary if the request is an HTMX request. I couldn't reproduce the issue in the test until I added the Htmx specific Http header to the request. I have updated the test to illustrate the issue.

Reply:
Extremely strange. Good luck debugging that one!

Reply:
False alarm with the Htmx stuff - embarassing, but phew... 

I still can't reproduce the CIV creation in the test though. In the test the form immediately evaluates to invalid (as it should) whereas in the UI if I click my way through adding a new interface with an incorrect value, the form first evaluates to true (and hence hits the `form_valid` method on the view) and only gets marked as invalid after another round of validation, triggered by the creation of the CIV. It's strange that the form initially evaluates to valid, it shouldn't in my eyes (see snippet that Mike added): 

https://github.com/comic/grand-challenge.org/blob/ea980be59a473e8cf225a458d4db6b9caa2a5243/app/grandchallenge/reader_studies/views.py#L1644-L1653

**In this snippet, since we catch the validation error, the transaction isn't rolled back and so we end up with a CIV without a value.** 

Does this make any sense? I have looked at this for a bit too long in a row today, and will do something else for the rest of today to regain some clarity.

Reply:
OK that makes more sense. Then maybe you need to do things the other way around - do not do `CIV.create(...)`, instead create the CIV directly then do the validation and if it passes save it.

Reply:
And yes, the validation should definitely be completed before form_valid is hit! It's no good adding errors to the form afterwards.

Reply:
Figured it out, I think! The interface field was not required, although it should be on this form. So empty values passed validation on the form's clean method, but failed validation on CIV creation. I suppose that Mike added the try... except statement in the form_valid method to catch this scenario and return the error to the user. If we make the field required, we no longer need that though. All other errors will then be caught in the form's clean method, or for the more complex cases in the asynchronous tasks (where we use atomic blocks everywhere to ensure that the transaction gets reverted if validation fails).